universe/ShipDesign.h

```diff
@@ -90,7 +94,10 @@ class FO_COMMON_API PartType {
     PartType(ShipPartClass part_class, double capacity, double stat2,
              CommonParams& common_params, const MoreCommonParams& more_common_params,
              std::vector<ShipSlotType> mountable_slot_types,
-             const std::string& icon, bool add_standard_capacity_effect = true);
+             const std::string& icon, bool add_standard_capacity_effect = true,
+             int precision = 2,
```

Comment:
Is a precision value necessary? There are random pick and priority sort conditions that would allow more flexible control than whatever hardcoded precision mechanisms you implement...

Reply:
You mean picking and priority via FOCS? I do not think its a good idea to put the targetting algorithm distributed into FOCS parts.

The precision is there to supply weights for e.g. priority sort. I dont like the "hardcoded" fighting mechanism we have, I would prefer that it would be scripted in python and pluggable.

Reply:
I am not sure I understand your point regarding FOCS. I would have used weighted conditions if such existed. I had a lot of trouble with the parser already, so that would have been a stretch goal. 

I wouldnt mind if somebody moved weights to FOCS conditions. 
I have the feeling that I would mind if the exact algorithm is specified in ship parts. I think the battle system should be pluggable, so the interpretation of the values could be different from one battle system to the next. The interpretation should not happen in the contents like the parts. 

Precision is used as in a double role at the moment. Higher precision gives a weight for target selection if there a multiple conditions and also increases the chance that a condition actually locks a target. I expect precision to express a level of actually getting a lock on a target (and e.g. a good tactical engine could contribute to this) and maybe preferences weights move to its own specification. If you have a better suggestion, i will refactor the naming of course (its the best i could come up with). 

I think the design is reasonable enough even with its debatable choices to be included in master (at least i hope so). 

For design changes/expansion we should move this discussion to the forum i think. For example i think it would be a good idea to separate precision from condition weights. But that is definitly out of scope for my implementation. I think it is time to clean up the code and put the design into the wild and then improve.

Reply:
The combat system can be "pluggable", but the way a part picks its targets is part of the design of that part, and not something that should take a few vaguely-defined values from the part and interpret in different ways.

I would drop the whole precision concept for this iteration. The point should just be to have a way to script what targets a part can attack, and apply that condition if present, to pick a part's allowed targets.

For the next iteration, you don't really need "weighted conditions". If species and part type both need to have input on target selection, then they can do so in a fixed order... eg. If there is a part targeting condition, filter the potential targets with that condition. Then, if there is a species targeting condition, apply that. For the latter, assuming a primary application would be prioritizing by species of the target, a sorting condition can be used, which checks for the species of the target and selects those targets first.